ql/src/java/org/apache/hadoop/hive/ql/Driver.java

```diff
@@ -149,207 +149,64 @@ private CommandProcessorResponse run(String command, boolean alreadyCompiled) th
-       * In practice that is rarely the case, so the messy logic below tries to tease
-       * out canonical error code if it can.  Exclude stack trace from output when
-       * the error is a specific/expected one.
-       * It's written to stdout for backward compatibility (WebHCat consumes it).*/
-      try {
-        if (cpe.getCause() == null) {
-          mdf.error(ss.out, cpe.getMessage(), cpe.getResponseCode(), cpe.getSqlState());
-          throw cpe;
-        }
-        ErrorMsg canonicalErr = ErrorMsg.getErrorMsg(cpe.getResponseCode());
-        if (canonicalErr != null && canonicalErr != ErrorMsg.GENERIC_ERROR) {
-          /*Some HiveExceptions (e.g. SemanticException) don't set
-            canonical ErrorMsg explicitly, but there is logic
-            (e.g. #compile()) to find an appropriate canonical error and
-            return its code as error code. In this case we want to
-            preserve it for downstream code to interpret*/
-          mdf.error(ss.out, cpe.getMessage(), cpe.getResponseCode(), cpe.getSqlState(), null);
-          throw cpe;
-        }
-        if (cpe.getCause() instanceof HiveException) {
-          HiveException rc = (HiveException)cpe.getCause();
-          mdf.error(ss.out, cpe.getMessage(), rc.getCanonicalErrorMsg().getErrorCode(), cpe.getSqlState(),
-              rc.getCanonicalErrorMsg() == ErrorMsg.GENERIC_ERROR ? StringUtils.stringifyException(rc) : null);
-        } else {
-          ErrorMsg canonicalMsg = ErrorMsg.getErrorMsg(cpe.getCause().getMessage());
-          mdf.error(ss.out, cpe.getMessage(), canonicalMsg.getErrorCode(), cpe.getSqlState(),
-              StringUtils.stringifyException(cpe.getCause()));
-        }
-      } catch (HiveException ex) {
-        CONSOLE.printError("Unable to JSON-encode the error", StringUtils.stringifyException(ex));
-      }
+      processRunException(cpe);
       throw cpe;
     }
   }
 
   private void runInternal(String command, boolean alreadyCompiled) throws CommandProcessorException {
     DriverState.setDriverState(driverState);
 
-    driverState.lock();
-    try {
-      if (driverContext != null && driverContext.getPlan() != null
-          && driverContext.getPlan().isPrepareQuery()
-          && !driverContext.getPlan().isExplain()) {
-        LOG.info("Skip running tasks for prepare plan");
-        return;
-      }
-      if (alreadyCompiled) {
-        if (driverState.isCompiled()) {
-          driverState.executing();
-        } else {
-          String errorMessage = "FAILED: Precompiled query has been cancelled or closed.";
-          CONSOLE.printError(errorMessage);
-          throw DriverUtils.createProcessorException(driverContext, 12, errorMessage, null, null);
-        }
-      } else {
-        driverState.compiling();
-      }
-    } finally {
-      driverState.unlock();
+    if (driverContext != null && driverContext.getPlan() != null &&
```

Comment:
This is nice because it has been moved outside the lock scope.  Do note however that driverContext can never be `null`.  It is defined as `final` and set in the constructor.  Please remove this check.

Reply:
I do see that the `driverContext` plan is set in a few different places.  I'm not sure if this can be checked outside of a lock.  At the very least, I would minimize the risk a bit:

```
QueryPlan plan = driverContext.getPlan();
if (plan != null
          && plan.isPrepareQuery()
          && !plan.isExplain()) {
```

At least you know that the value of `plan` cannot possibly change between calls.

Reply:
Agree, nice catch, fixed.